File: LeetCode/23. Merge k Sorted Lists/Solution.cc
#include <algorithm> // for std::min
#include <queue>
#include <vector>

using std::vector;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode *mergeKLists(vector<ListNode *> &lists) {
    auto cmp = [](const ListNode *lhs, const ListNode *rhs) { 
        return lhs->val > rhs->val;
    };
    std::priority_queue<ListNode *, std::vector<ListNode *>, decltype(cmp)> queue;
    ListNode hh;
    auto head = &hh;
    auto current = head;

    std::for_each(std::begin(lists), std::end(lists), [&](ListNode *n) {
      if (n) {
        queue.push(n);
      }
    });

    while (!queue.empty()) {
      auto node = queue.top();
      queue.pop();

      current->next = node;
      current = node;

      if (node->next) {
        queue.push(node->next);
      }
    }

    return head->next;
  }
};


int main() {}

File: LeetCode/23. Merge k Sorted Lists/Solution.kt
package problem23

import java.util.PriorityQueue

/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
  fun mergeKLists(lists: Array<ListNode?>): ListNode? {
    val queue = PriorityQueue<ListNode> { lhs, rhs ->
      lhs.`val` - rhs.`val`
    }

    repeat(lists.size) {
      lists[it]?.let { queue.offer(it) }
    }

    val head = ListNode(0)
    var current = head

    while (queue.isNotEmpty()) {
      val node = queue.poll()

      current.next = node
      current = node

      node.next?.let { queue.offer(it) }
    }

    return head.next
  }
}

class ListNode(var `val`: Int) {
  var next: ListNode? = null
}

// test

fun main() {
  // Solution().mergeKLists(arrayOf(
  //   intArrayOf(1, 4, 5),
  //   intArrayOf(1, 3, 4),
  //   intArrayOf(2, 6)
  // ))
  // Input: lists = [[1,4,5],[1,3,4],[2,6]] Output: [1,1,2,3,4,4,5,6]
  Solution().mergeKLists(emptyArray())
}